File: app-server-webflux/src/main/java/com/github/arhor/simple/expense/tracker/exception/PropertyConditionException.java
package com.github.arhor.simple.expense.tracker.exception;

import lombok.Getter;
import lombok.RequiredArgsConstructor;

@Getter
@RequiredArgsConstructor
public abstract class PropertyConditionException extends RuntimeException {

    private final String name;
    private final String condition;

    public Object[] getParams() {
        return new Object[]{name, condition};
    }
}

File: app-server/src/main/kotlin/com/github/arhor/simple/expense/tracker/exception/PropertyConditionException.kt
package com.github.arhor.simple.expense.tracker.exception;

abstract class PropertyConditionException(
    private val name: String,
    private val condition: String,
) : RuntimeException("Property [$name] exception occurred, caused by condition [$condition]"), Destructurable {

    override val components: Array<out Any?>
        get() = arrayOf(name, condition)
}

class EntityNotFoundException(name: String, condition: String) : PropertyConditionException(name, condition)
class EntityDuplicateException(name: String, condition: String) : PropertyConditionException(name, condition)
